### ISSUE TITLE
Performance / Caching

### ISSUE THREAD

**Issue by ruudk** [twigstan/twigstan#8]
PHPStan's cache misses all the time, should be fixed to speed things up significantly

**Comment by raffaelecarelle** [comment]
Do you think you should set phpstan's tmpDir option to the project directory or take the one set by the end user? Or would compiling the twig files invalidate the cache each time?

**Comment by ruudk** [comment]
We could do this in small steps.

First step is to use`.twigstan` (as a default) for the `tmpDir` in the current working directory. This is already done in the EndToEnd tests.

Once that is done, we should think about this:
https://github.com/twigstan/twigstan/blob/ffb40722b857e5b45a934547815b96303baf179f/src/Application/AnalyzeCommand.php#L167-L177

Here, we clear the directory before each run. That should not be done anymore.

But this leads to another issue: we currently let PHPStan collect the scope for all the found php files + the flattening temp directory: https://github.com/twigstan/twigstan/blob/ffb40722b857e5b45a934547815b96303baf179f/src/Application/AnalyzeCommand.php#L265

This means that if there are files in that directory, that no longer exist in the project, we still process them. We should make sure that those files are no longer there, otherwise PHPStan will do work for nothing.

There are more optimizations that can be done: if we already have a compiled Twig template, maybe there is no need to recompile it again next time.

**Comment by ruudk** [comment]
The filenames are already predictable. That means that a given path to a Twig template, always results in a single file with the path hashed:
https://github.com/twigstan/twigstan/blob/ffb40722b857e5b45a934547815b96303baf179f/src/Processing/Compilation/TwigCompiler.php#L92-L96

**Comment by raffaelecarelle** [comment]
hello @ruudk, I ran twigstan on whole my project (80 files found) and twigstan stucks on find dependencies for a lot of minutes without finish. 

If you agree I could start working on performance and caching improvement.

**Comment by ruudk** [comment]
Interesting. For now, I don't think it's good to work on this yet. There are a few things that I'm working on currently: php configs, simplifying the compiled code by keeping the original Twig $context arrays instead of the variables, rethinking the way we analyze templates etc. 

But maybe you can find out why the dependencies cannot be resolved? I have the feeling it's an endless loop somewhere. Would be great to distill this in a failing functional test.

**Comment by raffaelecarelle** [comment]
ok good, thank you
